runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -1564,11 +1568,25 @@ TR::Register *J9::X86::TreeEvaluator::multianewArrayEvaluator(TR::Node *node, TR
    bool use64BitClasses = comp->target().is64Bit() && !TR::Compiler->om.generateCompressedObjectHeaders();
    generateMemRegInstruction(TR::InstOpCode::SMemReg(use64BitClasses), node, generateX86MemoryReference(targetReg, TR::Compiler->om.offsetOfObjectVftField(), cg), classReg, cg);
 
-   // Init size and '0' fields to 0
+   // Init size and mustBeZero ('0') fields to 0
    generateMemImmInstruction(TR::InstOpCode::S4MemImm4, node, generateX86MemoryReference(targetReg, fej9->getOffsetOfContiguousArraySizeField(), cg), 0, cg);
    generateMemImmInstruction(TR::InstOpCode::S4MemImm4, node, generateX86MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousArraySizeField(), cg), 0, cg);
 
+#ifdef TR_TARGET_64BIT
+   // Load dataAddr slot offset difference since 0 size arrays are treated as discontiguous.
+   TR_ASSERT_FATAL_WITH_NODE(node
+      , IS_32BIT_SIGNED(fej9->getOffsetOfDiscontiguousDataAddrField() - fej9->getOffsetOfContiguousDataAddrField())
+      , "dataAddrFieldOffset is too big for the instruction.");
+
+   generateRegImm64Instruction(TR::InstOpCode::MOV8RegImm64
+      , node, temp3Reg
+      , static_cast<int32_t>(fej9->getOffsetOfDiscontiguousDataAddrField() - fej9->getOffsetOfContiguousDataAddrField())
+      , cg);
```

Comment:
Since you have an assert above guarding the range of this displacement, this instruction should be:
```
generateRegImmInstruction(TR::InstOpCode::MOV8RegImm4,
                          node, 
                          temp3Reg,
                          static_cast<int32_t>(fej9->getOffsetOfDiscontiguousDataAddrField() - fej9->getOffsetOfContiguousDataAddrField()),
                          cg);
```
Otherwise, you're still generating the big 10 byte instruction.